packages/loopring_v3.js/src/SparseMerkleTree.ts

```diff
@@ -0,0 +1,125 @@
+import poseidon = require("./poseidon");
```

Comment:
We probably don't need to construct any tree when we process the onchain data. So the tree will be flattened to its' leaves, so we'll have a table for balance, a table for trade history. Only when a user requests to construct the Merkle root, then we calculate on demand.

Reply:
The Merkle tree is only built when requested using the API (`buildMerkleTree` or `buildMerkleTreeForWithdrawalMode`). The Merkle tree does NOT get build incremental while the blocks are being processed and the leaf data is being updated.

Reply:
i see. this is also my understanding